src/laser_proc.cpp

```diff
@@ -27,20 +27,146 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-/* 
+/*
  * Author: Chad Rockey
  */
 
-#include <laser_proc/LaserProcROS.h>
+#include <limits>
+#include <vector>
+
+#include "laser_proc/laser_proc.hpp"
+
+using namespace laser_proc;  // NOLINT
+
+sensor_msgs::msg::LaserScan LaserProc::getFirstScan(
+  const sensor_msgs::msg::MultiEchoLaserScan & msg)
+{
+  sensor_msgs::msg::LaserScan out;
+  fillLaserScan(msg, out);
+  out.ranges.resize(msg.ranges.size());
+  if (msg.ranges.size() == msg.intensities.size()) {
+    out.intensities.resize(msg.intensities.size());
+  }
+
+  for (size_t i = 0; i < out.ranges.size(); i++) {
+    size_t index = getFirstValue(msg.ranges[i], out.ranges[i]);
+    if (out.intensities.size() > 0) {
+      if (msg.intensities[i].echoes.size() > 0) {
+        out.intensities[i] = msg.intensities[i].echoes[index];
+      } else {
+        out.intensities[i] = 0.0;
+      }
+    }
+  }
+  return out;
+}
+
+sensor_msgs::msg::LaserScan LaserProc::getLastScan(const sensor_msgs::msg::MultiEchoLaserScan & msg)
+{
+  sensor_msgs::msg::LaserScan out;
+  fillLaserScan(msg, out);
+  out.ranges.resize(msg.ranges.size());
+  if (msg.ranges.size() == msg.intensities.size()) {
+    out.intensities.resize(msg.intensities.size());
+  }
+  for (size_t i = 0; i < out.ranges.size(); i++) {
+    size_t index = getLastValue(msg.ranges[i], out.ranges[i]);
+    if (out.intensities.size() > 0) {
+      if (msg.intensities[i].echoes.size() > 0) {
+        out.intensities[i] = msg.intensities[i].echoes[index];
+      } else {
+        out.intensities[i] = 0.0;
+      }
+    }
+  }
+  return out;
+}
+
+sensor_msgs::msg::LaserScan LaserProc::getMostIntenseScan(
+  const sensor_msgs::msg::MultiEchoLaserScan & msg)
+{
+  sensor_msgs::msg::LaserScan out;
+  fillLaserScan(msg, out);
+  if (msg.ranges.size() == msg.intensities.size()) {
+    out.ranges.resize(msg.ranges.size());
+    out.intensities.resize(msg.intensities.size());
+  } else {
+    std::stringstream ss;
+    ss <<
+      "getMostIntenseScan::Size of ranges does not equal size of intensities, cannot create scan.";
+    throw std::runtime_error(ss.str());
```

Comment:
Throwing an exception here means that an external process can cause us to crash just by sending a malformed message.  This should be a warning and a return rather than an exception.

Reply:
This comment is still open.

Reply:
I was hesitating about introducing rclcpp here, but in the end opted for `RCLCPP_ERROR`. Let me know if a simple `fprintf` is preferred